Caching wrapper around scoring/threshold based filters

jelmervdl opened the pull request:
Initial implementation of the idea from https://github.com/jelmervdl/empty-train/issues/29.

Basically it rewrites the more expensive score-based filtering filters into something that just generates a score. Then the `threshold.py` program uses this score to determine which lines to filter.

The upside of all this is that the scores can be cached based on the line hash! So especially for the sample which is constantly being re-filtered, this should speed things up after the initial run.

Maybe in the future empty-train could show you the scores as an extra column. It also makes the filtering programs a bit simpler: they don't need to do the thresholding anymore. It will also be easier to add filters that filter rows based on (the score of) a single column.

Edit credit where credit's due: this was partially inspired by how [OpusFilter](https://github.com/Helsinki-NLP/OpusFilter) filters work: they each implement a score & and a threshold function. See https://github.com/Helsinki-NLP/OpusFilter/blob/develop/opusfilter/filters.py.

XapaJIaMnu commented:
The cache is per language, reopened for every file, right?

jelmervdl commented:
It runs on the stdin/stdout stream, it has no concept of individual files. The hash of the input line is the only thing that determines a cache hit/miss. The one filter that I use it for does LASER based similarity of both sides of a pair. So the hash & score is for the entire pair.

In that configuration file I've also made sure that different parameters for the filter will use a different cache:
https://github.com/jelmervdl/empty-train/pull/33/files#diff-500a4ae8e746e1059ec8a9cfc62e63b6e0bfccab17471e977d1e79c35ab9e557L30-R24

I thought about hard-coding it in such a way that any change in the command passed to threshold.py would result in a different cache, but then that also always include options like `--threads 4` or something, options that do not affect the output. So now it is up to the programmer.

I also thought about using a different cache based on the previous steps that come before the filter. But since we're scoring on the hash of the sentence pair, if any bit of it has been altered by a previous filter it will re-evaluate the pair anyway. And many pairs will stay the same if you change a threshold in a step before this one. So in the end it doesn't take any of the previous steps into account.

XapaJIaMnu commented on filters/laser_similarity.json:
```diff
@@ -27,5 +21,5 @@
         }
     },
-    "command": "./laser_similarity.py --threshold $THRESHOLD --batch-size $BATCHSIZE --src-lang $SRCLANG --tgt-lang $TGTLANG"
+    "command": "../threshold.py --cache $TMPDIR/laser.$(echo \"$SRCLANG $TGTLANG\" | cksum | cut -d' ' -f1).dbm $THRESHOLD ./laser_similarity.py --scores --src-lang $SRCLANG --tgt-lang $TGTLANG"
```
How does this run? I see no `__main__()` or anything? Where does python default to?

XapaJIaMnu reviewed (commented):
The reason I am asking is, because I can't actually run it, because I can't install pyhash. 
```
Collecting pyhash
  Using cached pyhash-0.9.3.tar.gz (602 kB)
  Preparing metadata (setup.py) ... error
  error: subprocess-exited-with-error
  
  × python setup.py egg_info did not run successfully.
  │ exit code: 1
  ╰─> [54 lines of output]
      /home/dheart/uni_stuff/postdoc/empty-train/.env/lib/python3.10/site-packages/setuptools/config/setupcfg.py:463: SetuptoolsDeprecationWarning: The license_file parameter is deprecated, use license_files instead.
        warnings.warn(msg, warning_class)
      /home/dheart/uni_stuff/postdoc/empty-train/.env/lib/python3.10/site-packages/setuptools/installer.py:27: SetuptoolsDeprecationWarning: setuptools.installer is deprecated. Requirements should be satisfied by a PEP 517 installer.
        warnings.warn(
      WARNING: The wheel package is not available.
      WARNING: The wheel package is not available.
      WARNING: The wheel package is not available.
        error: subprocess-exited-with-error
      
        × python setup.py bdist_wheel did not run successfully.
        │ exit code: 1
        ╰─> [6 lines of output]
            usage: setup.py [global_opts] cmd1 [cmd1_opts] [cmd2 [cmd2_opts] ...]
               or: setup.py --help [cmd1 cmd2 ...]
               or: setup.py --help-commands
               or: setup.py cmd --help
      
            error: invalid command 'bdist_wheel'
            [end of output]
      
        note: This error originates from a subprocess, and is likely not a problem with pip.
        ERROR: Failed building wheel for py-cpuinfo
      ERROR: Failed to build one or more wheels
      Traceback (most recent call last):
        File "/home/dheart/uni_stuff/postdoc/empty-train/.env/lib/python3.10/site-packages/setuptools/installer.py", line 82, in fetch_build_egg
          subprocess.check_call(cmd)
        File "/usr/lib/python3.10/subprocess.py", line 369, in check_call
          raise CalledProcessError(retcode, cmd)
      subprocess.CalledProcessError: Command '['/home/dheart/uni_stuff/postdoc/empty-train/.env/bin/python3', '-m', 'pip', '--disable-pip-version-check', 'wheel', '--no-deps', '-w', '/tmp/tmpxz79gruj', '--quiet', 'py-cpuinfo']' returned non-zero exit status 1.
      
      The above exception was the direct cause of the following exception:
      
      Traceback (most recent call last):
        File "<string>", line 2, in <module>
        File "<pip-setuptools-caller>", line 34, in <module>
        File "/tmp/pip-install-srvlwu_d/pyhash_630c67813d8f487aa64ac41b4cae8390/setup.py", line 129, in <module>
          setup(name='pyhash',
        File "/home/dheart/uni_stuff/postdoc/empty-train/.env/lib/python3.10/site-packages/setuptools/__init__.py", line 86, in setup
          _install_setup_requires(attrs)
        File "/home/dheart/uni_stuff/postdoc/empty-train/.env/lib/python3.10/site-packages/setuptools/__init__.py", line 80, in _install_setup_requires
          dist.fetch_build_eggs(dist.setup_requires)
        File "/home/dheart/uni_stuff/postdoc/empty-train/.env/lib/python3.10/site-packages/setuptools/dist.py", line 875, in fetch_build_eggs
          resolved_dists = pkg_resources.working_set.resolve(
        File "/home/dheart/uni_stuff/postdoc/empty-train/.env/lib/python3.10/site-packages/pkg_resources/__init__.py", line 789, in resolve
          dist = best[req.key] = env.best_match(
        File "/home/dheart/uni_stuff/postdoc/empty-train/.env/lib/python3.10/site-packages/pkg_resources/__init__.py", line 1075, in best_match
          return self.obtain(req, installer)
        File "/home/dheart/uni_stuff/postdoc/empty-train/.env/lib/python3.10/site-packages/pkg_resources/__init__.py", line 1087, in obtain
          return installer(requirement)
        File "/home/dheart/uni_stuff/postdoc/empty-train/.env/lib/python3.10/site-packages/setuptools/dist.py", line 945, in fetch_build_egg
          return fetch_build_egg(self, req)
        File "/home/dheart/uni_stuff/postdoc/empty-train/.env/lib/python3.10/site-packages/setuptools/installer.py", line 84, in fetch_build_egg
          raise DistutilsError(str(e)) from e
      distutils.errors.DistutilsError: Command '['/home/dheart/uni_stuff/postdoc/empty-train/.env/bin/python3', '-m', 'pip', '--disable-pip-version-check', 'wheel', '--no-deps', '-w', '/tmp/tmpxz79gruj', '--quiet', 'py-cpuinfo']' returned non-zero exit status 1.
      [end of output]
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
error: metadata-generation-failed

× Encountered error while generating package metadata.
╰─> See above for output.

note: This is an issue with the package mentioned above, not pip.
hint: See above for details.
```
Pypy seems to mark the build as broken: 
It seems to require to downgrade setuptools. Are we ok doing that in `requirements.txt`?
https://github.com/flier/pyfasthash/issues/62

jelmervdl commented:
Oh yes I had forgotten about that but already (suppressing bad memories I guess) but I'm very open to other stable and fast hashing functions/libraries. This is just the one I see most commonly used.

XapaJIaMnu commented:
How about this one? https://github.com/ifduyue/python-xxhash

XapaJIaMnu commented on threshold.py:
```diff
@@ -0,0 +1,237 @@
+import argparse
+import pickle
+import dbm
+from pyhash import murmur3_32
```
```python
from xxhash import xxh64
```

XapaJIaMnu commented on threshold.py:
```diff
@@ -0,0 +1,237 @@
+	the queue and the order of feeding fchild are the same, the
+	`threshold_scores` thread will know how to link them back together.
+	"""
+	derive_key = murmur3_32()
```
remove

XapaJIaMnu reviewed (commented):
I've added this as a suggestion as I wasn't sure exactly how you test it so better leave testing to you. Also:
Add
```python
xxhash==3.0.0
```
To requirements.txt

jelmervdl commented:
I've changed it to xxhash! I also updated run.py and col.py to treat SIGPIPE, the signal that happens when one side of the pipe closes, happens. This pretty much works nicely when you put `head -n X` before, after, or in the middle of a pipeline. It still fails properly if a different type of error occurs.

XapaJIaMnu commented on threshold.py:
```diff
@@ -0,0 +1,265 @@
+	the queue and the order of feeding fchild are the same, the
+	`threshold_scores` thread will know how to link them back together.
+	"""
+	derive_key = lambda val: xxh32(val).digest()
```
it defaults to the same seed on every system?
